sklearn/tests/test_metadata_routing.py

```diff
@@ -70,10 +70,18 @@ def assert_request_equal(request, dictionary):
         assert not len(getattr(request, method).requests)
 
 
-def record_metadata(obj, method, **kwargs):
-    """Utility function to store passed metadata to a method."""
+def record_metadata(obj, method, record_none=True, **kwargs):
+    """Utility function to store passed metadata to a method.
+
+    If record_none is False, kwargs whose values are None are skipped. This is
+    so that checks on keyword arguments whose default was not changed are
+    skipped.
```

Comment:
We record and check None to make sure if a metadata is not requested, it is not passed, not even as None.

The tests which don't record None, work because the user is not passing any metadata as None, and we need to ignore them cause the default value is `None` and explicitly set in those sub-estimator methods.

Alternatively, we could change the default to `"default"`, and let `record_metadata` know what the _default_ value is, and only ignore that. It might be cleaner.

Reply:
What do you think about always recording `None` and have a flag in `check_recorded_metadata` to switch between the two modes of checking? Concretely:

```python
def check_recorded_metadata(obj, method, strict=True, **kwargs):
    """Check whether the expected metadata is passed to the object's method."""
    records = getattr(obj, "_records", dict()).get(method, dict())

    cmp = operator.eq if strict else operator.le
    assert cmp(set(kwargs.keys()), set(records.keys()))
    for key, value in kwargs.items():
        assert records[key] is value
```

Reply:
I kinda prefer the current implementation because it's easier for the test to test exactly what it needs to. What you have here kind of a subset of what the current implementation does, as in, there's no way to test if the router has routed None explicitly or not, and that's something we need to check. I might be missing something here though.